File: internal/agent/ports/agent/context.go
package agent

import (
	"context"
	"time"

	core "alex/internal/agent/ports"
	"alex/internal/agent/ports/storage"
)

// ContextManager handles layered context orchestration across static, dynamic and
// meta layers.
type ContextManager interface {
	// EstimateTokens estimates token count for messages.
	EstimateTokens(messages []core.Message) int

	// Compress reduces message size when limit approached.
	Compress(messages []core.Message, targetTokens int) ([]core.Message, error)

	// AutoCompact applies compression automatically when the configured
	// threshold is exceeded. It returns the possibly compacted slice and a flag
	// indicating whether compaction occurred.
	AutoCompact(messages []core.Message, limit int) ([]core.Message, bool)

	// ShouldCompress checks if compression needed.
	ShouldCompress(messages []core.Message, limit int) bool

	// Preload ensures the manager has cached static context/configuration before
	// first use.
	Preload(ctx context.Context) error

	// BuildWindow composes the full context window for the given session and
	// configuration.
	BuildWindow(ctx context.Context, session *storage.Session, cfg ContextWindowConfig) (ContextWindow, error)

	// RecordTurn writes the supplied turn record to the dynamic state store so
	// that UI/API consumers can replay the session.
	RecordTurn(ctx context.Context, record ContextTurnRecord) error
}

// SessionContextKey is the shared context key for storing session IDs across packages.
// This ensures consistent session ID propagation from server layer to agent layer.
type SessionContextKey struct{}

// ContextWindowConfig drives context composition behaviour.
type ContextWindowConfig struct {
	TokenLimit         int
	PersonaKey         string
	GoalKey            string
	WorldKey           string
	ToolMode           string
	ToolPreset         string
	EnvironmentSummary string
}

// ContextWindow exposes the layered context returned by the manager.
type ContextWindow struct {
	SessionID    string         `json:"session_id"`
	Messages     []core.Message `json:"messages"`
	SystemPrompt string         `json:"system_prompt"`
	Static       StaticContext  `json:"static"`
	Dynamic      DynamicContext `json:"dynamic"`
	Meta         MetaContext    `json:"meta"`
}

// ContextWindowPreview bundles the constructed window with metadata useful for
// debugging and visualization.
type ContextWindowPreview struct {
	Window        ContextWindow `json:"window"`
	TokenEstimate int           `json:"token_estimate"`
	TokenLimit    int           `json:"token_limit"`
	PersonaKey    string        `json:"persona_key,omitempty"`
	ToolMode      string        `json:"tool_mode,omitempty"`
	ToolPreset    string        `json:"tool_preset,omitempty"`
}

// StaticContext captures persona, goals, rules and knowledge packs.
type StaticContext struct {
	Persona            PersonaProfile       `json:"persona"`
	Goal               GoalProfile          `json:"goal"`
	Policies           []PolicyRule         `json:"policies"`
	Knowledge          []KnowledgeReference `json:"knowledge"`
	Tools              []string             `json:"tools"`
	World              WorldProfile         `json:"world"`
	UserPersona        *core.UserPersonaProfile `json:"user_persona,omitempty"`
	EnvironmentSummary string               `json:"environment_summary,omitempty"`
	Version            string               `json:"version,omitempty"`
}

// DynamicContext contains plans, beliefs and world-state diffs.
type DynamicContext struct {
	TurnID            int              `json:"turn_id"`
	LLMTurnSeq        int              `json:"llm_turn_seq"`
	Plans             []PlanNode       `json:"plans"`
	Beliefs           []Belief         `json:"beliefs"`
	WorldState        map[string]any   `json:"world_state"`
	Feedback          []FeedbackSignal `json:"feedback"`
	SnapshotTimestamp time.Time        `json:"snapshot_timestamp"`
}

// MetaContext tracks long-horizon memories and persona bookkeeping.
type MetaContext struct {
	Memories        []MemoryFragment `json:"memories"`
	Recommendations []string         `json:"recommendations"`
	PersonaVersion  string           `json:"persona_version"`
}

// PersonaProfile models persona level instructions.
type PersonaProfile struct {
	ID            string `json:"id" yaml:"id"`
	Tone          string `json:"tone" yaml:"tone"`
	RiskProfile   string `json:"risk_profile" yaml:"risk_profile"`
	DecisionStyle string `json:"decision_style" yaml:"decision_style"`
	Voice         string `json:"voice" yaml:"voice"`
}

// GoalProfile enumerates long and mid-term goals.
type GoalProfile struct {
	ID             string   `json:"id" yaml:"id"`
	LongTerm       []string `json:"long_term" yaml:"long_term"`
	MidTerm        []string `json:"mid_term" yaml:"mid_term"`
	SuccessMetrics []string `json:"success_metrics" yaml:"success_metrics"`
}

// PolicyRule contains explicit guardrails/preference statements.
type PolicyRule struct {
	ID              string   `json:"id" yaml:"id"`
	HardConstraints []string `json:"hard_constraints" yaml:"hard_constraints"`
	SoftPreferences []string `json:"soft_preferences" yaml:"soft_preferences"`
	RewardHooks     []string `json:"reward_hooks" yaml:"reward_hooks"`
}

// KnowledgeReference references SOP or RAG collections.
type KnowledgeReference struct {
	ID             string   `json:"id" yaml:"id"`
	Description    string   `json:"description" yaml:"description"`
	SOPRefs        []string `json:"sop_refs" yaml:"sop_refs"`
	RAGCollections []string `json:"rag_collections" yaml:"rag_collections"`
	MemoryKeys     []string `json:"memory_keys" yaml:"memory_keys"`

	// ResolvedSOPContent holds the resolved markdown content for each SOP ref.
	// Populated at runtime by SOPResolver; never read from config.
	ResolvedSOPContent map[string]string `json:"resolved_sop_content,omitempty" yaml:"-"`
}

// WorldProfile enumerates runtime environment capabilities and limits.
type WorldProfile struct {
	ID           string   `json:"id" yaml:"id"`
	Environment  string   `json:"environment" yaml:"environment"`
	Capabilities []string `json:"capabilities" yaml:"capabilities"`
	Limits       []string `json:"limits" yaml:"limits"`
	CostModel    []string `json:"cost_model" yaml:"cost_model"`
}

// PlanNode encodes nested plan trees.
type PlanNode struct {
	ID          string     `json:"id"`
	Title       string     `json:"title"`
	Status      string     `json:"status"`
	Children    []PlanNode `json:"children"`
	Description string     `json:"description,omitempty"`
}

// Belief captures assumptions or facts.
type Belief struct {
	Statement  string  `json:"statement"`
	Confidence float64 `json:"confidence"`
	Source     string  `json:"source,omitempty"`
}

// FeedbackSignal carries lightweight scoring signals.
type FeedbackSignal struct {
	Kind      string    `json:"kind"`
	Message   string    `json:"message"`
	Value     float64   `json:"value"`
	CreatedAt time.Time `json:"created_at"`
}

// MemoryFragment references retained memories.
type MemoryFragment struct {
	Key       string    `json:"key"`
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"created_at"`
	Source    string    `json:"source"`
}

// ContextTurnRecord is persisted each time the LLM is invoked.
type ContextTurnRecord struct {
	SessionID     string               `json:"session_id"`
	TurnID        int                  `json:"turn_id"`
	LLMTurnSeq    int                  `json:"llm_turn_seq"`
	Timestamp     time.Time            `json:"timestamp"`
	Summary       string               `json:"summary"`
	Plans         []PlanNode           `json:"plans"`
	Beliefs       []Belief             `json:"beliefs"`
	World         map[string]any       `json:"world_state"`
	Diff          map[string]any       `json:"diff"`
	Messages      []core.Message       `json:"messages"`
	Feedback      []FeedbackSignal     `json:"feedback"`
	KnowledgeRefs []KnowledgeReference `json:"knowledge_refs"`
}

File: web/lib/types/api/context.ts
import type { AttachmentPayload } from '../ui/attachment';
import type { UserPersonaProfile } from './persona';

export type MessageSource =
  | 'system_prompt'
  | 'user_input'
  | 'user_history'
  | 'assistant_reply'
  | 'tool_result'
  | 'debug'
  | 'evaluation';

export interface ToolCall {
  id: string;
  name: string;
  arguments: Record<string, any>;
  session_id?: string;
  task_id?: string;
  parent_task_id?: string;
}

export interface ToolResult {
  call_id: string;
  content: string;
  error?: string;
  metadata?: Record<string, any>;
  attachments?: Record<string, AttachmentPayload> | null;
}

export interface Message {
  role: string;
  content: string;
  tool_calls?: ToolCall[];
  tool_results?: ToolResult[];
  tool_call_id?: string;
  metadata?: Record<string, any>;
  attachments?: Record<string, AttachmentPayload> | null;
  source?: MessageSource;
}

export interface PersonaProfile {
  id?: string;
  tone?: string;
  risk_profile?: string;
  decision_style?: string;
  voice?: string;
}

export interface GoalProfile {
  id?: string;
  long_term?: string[];
  mid_term?: string[];
  success_metrics?: string[];
}

export interface PolicyRule {
  id?: string;
  hard_constraints?: string[];
  soft_preferences?: string[];
  reward_hooks?: string[];
}

export interface KnowledgeReference {
  id?: string;
  description?: string;
  sop_refs?: string[];
  memory_keys?: string[];
}

export interface WorldProfile {
  id?: string;
  environment?: string;
  capabilities?: string[];
  limits?: string[];
  cost_model?: string[];
}

export interface PlanNode {
  id: string;
  title: string;
  status: string;
  description?: string;
  children?: PlanNode[];
}

export interface Belief {
  statement: string;
  confidence?: number;
  source?: string;
}

export interface FeedbackSignal {
  kind: string;
  message: string;
  value?: number;
  created_at?: string;
}

export interface MemoryFragment {
  key: string;
  content: string;
  created_at?: string;
  source?: string;
}

export interface StaticContext {
  persona?: PersonaProfile;
  goal?: GoalProfile;
  policies?: PolicyRule[];
  knowledge?: KnowledgeReference[];
  tools?: string[];
  world?: WorldProfile;
  user_persona?: UserPersonaProfile;
  environment_summary?: string;
  version?: string;
}

export interface DynamicContext {
  turn_id?: number;
  llm_turn_seq?: number;
  plans?: PlanNode[];
  beliefs?: Belief[];
  world_state?: Record<string, any>;
  feedback?: FeedbackSignal[];
  snapshot_timestamp?: string;
}

export interface MetaContext {
  memories?: MemoryFragment[];
  recommendations?: string[];
  persona_version?: string;
}

export interface ContextWindow {
  session_id: string;
  messages: Message[];
  system_prompt?: string;
  static?: StaticContext;
  dynamic?: DynamicContext;
  meta?: MetaContext;
}

export interface ContextWindowPreviewResponse {
  session_id: string;
  token_estimate?: number;
  token_limit?: number;
  persona_key?: string;
  tool_mode?: string;
  tool_preset?: string;
  window: ContextWindow;
}
